### DISCUSSION TITLE
Eclipse Installation Manager and asking about Eclipse Oomph capabilities

### DISCUSSION THREAD

**Discussion by A7exSchin** [eclipse-oomph/oomph#17] [Ideas]
Hello there Oomph Team!

I want to use this discussion to ask some questions about the capabiltilities of the Oomph Project, more precisely the Setup Model and EMF. I also want to directly connect that to my current project which is my Bachelors Thesis.

I am currently developing the [Eclipse Installation Manager](https://github.com/A7exSchin/EclipseInstallationManager) which aims to provide the following functionalities

- Display installed installations with their corresponding workspaces
- Enable the developer to start installations with a specific workspace
- Provide a comprehensive overview over installations and workspaces
- Enable deletion and configuration f installations and workspaces

Currently I am strongly dependent on the Eclipse Oomph Setup Model:
I am using the `locations.setup` file the Eclipse Installer creates to find all installations (done with the installer) and mapped workspaces. I map them to a data structure. With this structure I can provide a (at least I think) valuable overviews:

The Tray Overview
![image](https://github.com/eclipse-oomph/oomph/assets/35499953/46bfd108-23c5-4a09-adad-d1ed03105292)

The Management Overview (not yet released)
![image](https://github.com/eclipse-oomph/oomph/assets/35499953/e3d34dc5-adc5-452a-94cf-27b0c4f2dfa4)

Now, I am facing the last capabilities:
Editing Installations, Deleting them and Configuring them. I am quite new to Eclipse, OSGi and EMF. I want to modify installations so they are consinstent with the Oomph Model. Therefore I have several questions:

1. Does the model provide anything with what I can delete installations and the needed resources get updates automatically?
2. When I last tried to edit eclipse.ini settings/proeprties, they got overwritten at start up. I read about an engine which keeps the installations matching their definition (Product Setup file). Does the engine update these parts if I update the Product Setup in the catalog? Can I use the Model to modify the eclipse.ini so they are consistent and are not begn overwritten?

Now I understand that this is probably way out of scope for what the Model has been planned. On the other hand, I really want this application to be as useful as possible. I actually like using it a lot. 

I have some more questions regarding updating installations and adding installations not done with the eclipse installer. I need to ask them because my Prof as suggested said features...

3. Can I add installations, that havent been done with the Eclipse Installer into the locations catalog? Can I "inject" Oomph capabilities into the installations? (I understand that with every installation done with the Eclipse Installer, parts of Oomph are installed alongside to keep the model in check. Thats what I want to recreate in unmanaged installations...)
4. This is a long shot and I do not think it is possible with the Oomph Model, but is there a capability with which I can trigger an Eclipse Update for an installation? A simple "no" should suffice, as I haven't found anything :)

Thanks for ANY help. I would also geatly appreciate any insights/help/opinions on my own project. I also would need some more people later down the line for an evaluation study. If you are interested message me or join the discussion on Github.

Thanks for the time! Again, I really appreciate it and love the concept of the Eclipse Installer/Oomph Project. The capabilities are insane.

**Comment by merks** [comment]
What you describe sound cool and useful.  I started something like that ages ago, but never got very far and then I was discouraged when I showed a prototype to someone "important" and their first reaction was, what is that useful for?  😱 

> I want to modify installations so they are consinstent with the Oomph Model. Therefore I have several questions:
> 
>     1. Does the model provide anything with what I can delete installations and the needed resources get updates automatically?
> 

No, there is nothing that will help delete installations or workspaces.  There are potentially useful utilities associated with Window -> Preferences -> Oomph -> Bundle Pools for analysis and cleanup.  These are also available from the installer by BUNDLE POOLS.../Manage Bundle Pools.

>     2. When I last tried to edit eclipse.ini settings/proeprties, they got overwritten at start up. I read about an engine which keeps the installations matching their definition (Product Setup file). Does the engine update these parts if I update the Product Setup in the catalog? Can I use the Model to modify the eclipse.ini so they are consistent and are not begn overwritten?
> 

There are EclipseIni tasks that will run and modify the eclipse.ini if they eclipse.ini doesn't contain the property as specified by the task.  So yes, by modifying such a task, it will performed on next startup (and generally a restart is required after modifying the eclipse.in..  There are also p2 touchpoint instructions that can and do modify the eclipse.ini when installing or updating content.

> 
> Now I understand that this is probably way out of scope for what the Model has been planned. On the other hand, I really want this application to be as useful as possible. I actually like using it a lot.
> 
> I have some more questions regarding updating installations and adding installations not done with the eclipse installer. I need to ask them because my Prof as suggested said features...
> 
>     3. Can I add installations, that havent been done with the Eclipse Installer into the locations catalog? 

It's been a while since I looked closely, but if Oomph setup is installed in any installation, I would expect it to add that installation (and workspace) to the locations setup.

> Can I "inject" Oomph capabilities into the installations? (I understand that with every installation done with the Eclipse Installer, parts of Oomph are installed alongside to keep the model in check. Thats what I want to recreate in unmanaged installations...)
> 

Primarily installing the setup feature into an installation should do the trick.    It should be possible to use p2's director application to add things to an existing installation, but I don't have good experience with using that directly...

>     4. This is a long shot and I do not think it is possible with the Oomph Model, but is there a capability with which I can trigger an Eclipse Update for an installation? A simple "no" should suffice, as I haven't found anything :)
> 

In principle this should be possible via the director application (which is effectively what the p2 director task is using).  It's just tricky to configure everything correctly using the correct agent, bundle pool and profile, and it's very easy to mix up the agent, pool, and profile of the installer itself with that of the installation with which it is work...

> 
> Thanks for ANY help. I would also geatly appreciate any insights/help/opinions on my own project. I also would need some more people later down the line for an evaluation study. If you are interested message me or join the discussion on Github.
> 

It definitely looks promising, not to mention cool and useful.

> Thanks for the time! Again, I really appreciate it and love the concept of the Eclipse Installer/Oomph Project. The capabilities are insane.

**Reply by A7exSchin** [reply]
Thanks for the reply!
I am sad to hear that higher ups did not think this wouldnt be useful. I use it basically every day and I think many would...

Do you by any chance have an idea how I can use the Bundle Tools / Cleanup Utilities from code? That would be really helpful

**Reply by merks** [reply]
Sometimes higher up also means not so directly involved in development and therefore the issue how how to manage dozens of installation isn't recognized as a problem.  I would need to look more deeply into the code to see what drives the analysis and cleanup logic.  Given it's open source, you can do that too.   😮

**Reply by A7exSchin** [reply]
I have been doing that for the LocationCatalog Impl. Just thought you might have a quick hint. I will take a look thanks!

**Reply by HannesWell** [reply]
> What you describe sound cool and useful. I started something like that ages ago, but never got very far and then I was discouraged when I showed a prototype to someone "important" and their first reaction was, what is that useful for? 😱

Is that person still important?

When Eclipse is compared with other IDEs the workspace selection dialog is one of the things that is mentioned as element that looks outdated and it's one of the first things visible to users.
And IIRC there are even external plugins to replace that.
So in general some modernization in that area would be nice and if we could even manage/select the installation to start, it would be even nicer. Then one just needs one icon on the desktop/task-bar instead of multiple.
